Add Trinidad and Tobago holidays

Roshan-1024 opened the pull request:
## Proposed change

Add Trinidad and Tobago holidays

Resolves #1289 .

## Type of change

- [x] New country/market holidays support (thank you!)
- [ ] Supported country/market holidays update (calendar discrepancy fix, localization)
- [ ] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency update (version deprecation/pin/upgrade)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [ ] New feature (new `holidays` functionality in general)

## Checklist

<!--
  Put an `x` in the boxes that apply. You can change them after PR is created.
-->

- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] I've successfully run `make check`, all checks and tests are green

<!--
  Thanks again for your contribution!
-->

[contributing-guidelines]: https://github.com/vacanza/holidays/blob/dev/CONTRIBUTING.md
[docs]: https://github.com/vacanza/holidays/tree/dev/docs/source

PPsyrius commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,33 @@
+from holidays.holiday_base import HolidayBase
+
+
+class TrinidadAndTobago(HolidayBase):
+    """
+    Trinidad and Tobago Holidays.
+
+    References:
+      * <https://en.wikipedia.org/wiki/Public_holidays_in_Trinidad_and_Tobago>
+      * <https://otp.tt/trinidad-and-tobago/national-holidays-and-awards/>
+    """
+
+    country = "TT"
+
```
@coderabbitai Consider suggesting newer examples from within the library i.e. `Macau.py`, `UnitedStates.py`, `Australia.py`

```
   def _populate_public_holidays(self):
        # New Year's Day
        self._add_new_years_day("New Year's Day")
```

PPsyrius commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,33 @@
+    country = "TT"
+
+
+class TT(TrinidadAndTobago):
+    pass
+
+
+class TTO(TrinidadAndTobago):
+    pass
```
@coderabbitai The current library-wide format doesn't add docstrings for two-letter and three-letter ISO 3166 country code aliases.

PPsyrius commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,33 @@
+      * <https://otp.tt/trinidad-and-tobago/national-holidays-and-awards/>
+    """
+
+    country = "TT"
```
@coderabbitai for English-speaking countries, i.e. Trinidad and Tobago - `default_language` and `supported_languages` parameter is entirely optional, but if languages are added for them, ensure that `en_US` is in the list alongside `en_[two letter regional code]`, i.e. `en_TT` in this case

PPsyrius commented on holidays/locale/en_US/LC_MESSAGES/TT.po:
```diff
@@ -0,0 +1,97 @@
+msgid "Indian Arrival Day"
+msgstr "Indian Arrival Day"
+
+#. Labour Day.
+msgid "Labour Day"
+msgstr "Labour Day"
```
```suggestion
#. Labour Day.
msgid "Labour Day"
msgstr "Labor Day"
```

PPsyrius commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,25 @@
+    def setUpClass(cls):
+        super().setUpClass(TrinidadAndTobago)
+
+    def test_country_aliases(self):
+        self.assertAliases(TrinidadAndTobago, TT, TTO)
```
This example is wrong, please don't listen to the bot

PPsyrius commented on holidays/locale/en_US/LC_MESSAGES/TT.po:
```diff
@@ -0,0 +1,97 @@
+# SOME DESCRIPTIVE TITLE
+# This file is distributed under the same license as the Holidays package.
+# FIRST AUTHOR <EMAIL@ADDRESS>, 2025.
+#, fuzzy
+msgid ""
+msgstr ""
```
Superceded by another human code reviewer comment

Roshan-1024 commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,186 @@
+
+
+class TrinidadAndTobagoIslamicHolidays(_CustomIslamicHolidays):
+    EID_AL_ADHA_DATES = {
+        2002: (FEB, 23),
+        2003: (FEB, 12),
+        2004: (FEB, 2),
+        2005: (JAN, 21),
+        2006: (JAN, 10),
+        2007: (DEC, 20),
+        2008: (DEC, 9),
+        2009: (NOV, 28),
+        2010: (NOV, 17),
+        2011: (NOV, 7),
+        2012: (OCT, 26),
+        2013: (OCT, 15),
+        2014: (OCT, 4),
+        2015: (SEP, 24),
+        2016: (SEP, 12),
+        2017: (SEP, 2),
+        2018: (AUG, 21),
+        2019: (AUG, 11),
+        2020: (JUL, 31),
+        2021: (JUL, 20),
+        2022: (JUL, 9),
+        2023: (JUN, 28),
+        2024: (JUN, 16),
+        2025: (JUN, 6),
+    }
+
```
You're right! thanks for pointing it out. While searching for `Eid al-Adha` holiday dates, the top results I found were for Saudi Arabia, and I mistakenly used those... I’ll remove this.

Roshan-1024 commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,158 @@
+        # Corpus Christi.
+        self._add_corpus_christi_day(tr("Corpus Christi"))
+
+        # Eid al-Fitr.
+        self._add_eid_al_fitr_day(tr("Eid-al-Fitr"))
```
Actually, `Eid al-Fitr` is written in [Wikipedia](https://en.wikipedia.org/wiki/Public_holidays_in_Trinidad_and_Tobago)

PPsyrius replied:
According to the [government source](https://otp.tt/trinidad-and-tobago/national-holidays-and-awards/) that we're using as a base for `en_TT` they uses "Eid-Ul-Fitr".

@coderabbitai the code comment should be in `en_US` so it's `Eid al-Fitr` - but the bot comment is correct about the main text here
```suggestion
        # Eid al-Fitr.
        self._add_eid_al_fitr_day(tr("Eid-Ul-Fitr"))
```

Roshan-1024 commented on holidays/locale/en_TT/LC_MESSAGES/TT.po:
```diff
@@ -0,0 +1,97 @@
+"Project-Id-Version: Holidays 0.70\n"
+"POT-Creation-Date: 2025-04-04 20:36+0530\n"
+"PO-Revision-Date: 2025-04-04 20:36+0530\n"
+"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
```
Do I need to add it in en_US/LC_MESSAGES/TT.po file as well?

PPsyrius replied:
Yes, you need to do this for all `.po` files 👍

PPsyrius replied:
Yes, you need to do this for all `.po` files 👍

PPsyrius commented on holidays/locale/en_US/LC_MESSAGES/TT.po:
```diff
@@ -0,0 +1,97 @@
+
+#. Eid al-Fitr.
+msgid "Eid-al-Fitr"
+msgstr "Eid-al-Fitr"
```
```suggestion
msgstr "Eid al-Fitr"
```

PPsyrius commented on holidays/locale/en_US/LC_MESSAGES/TT.po:
```diff
@@ -0,0 +1,97 @@
+
+#. Labor Day.
+msgid "Labour Day"
+msgstr "Labour Day"
```
```suggestion
msgstr "Labor Day"
```

PPsyrius commented on holidays/locale/en_TT/LC_MESSAGES/TT.po:
```diff
@@ -0,0 +1,96 @@
+msgid "Indian Arrival Day"
+msgstr ""
+
+#. Labor Day.
+msgid "Labour Day"
+msgstr ""
```
@coderabbitai Code comment is always in `en_US`, the `msgid` here instead `en_TT` which is the `default_language` for Trinidad and Tobago

KJhellico commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,158 @@
+        # Indian Arrival Day.
+        self._add_holiday_may_30(tr("Indian Arrival Day"))
+
+        # Labor Day.
+        self._add_holiday_jun_19(tr("Labour Day"))
```
@coderabbitai, it's en_US locale is standard for l10n comments.

PPsyrius commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,270 @@
+    @classmethod
+    def setUpClass(cls):
+        years = range(1963, 2050)
+        super().setUpClass(TrinidadAndTobago, years=years, years_non_observed=years)
```
For islamic holidays tests:
```suggestion
        super().setUpClass(TrinidadAndTobago, years=years, years_non_observed=years)
        cls.no_estimated_holidays = TrinidadAndTobago(years=years, islamic_show_estimated=False)
```

PPsyrius commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,270 @@
+            "2024-04-10",
+            "2025-03-31",
+        )
+        self.assertHolidayName(name, range(2012, 2025))
```
```suggestion
        self.assertHolidayName(name, self.no_estimated_holidays, range(1963, 2050))
```

PPsyrius commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,270 @@
+            ("2025-03-30", "Spiritual Baptist Shouter Liberation Day"),
+            ("2025-03-31", "Eid-Ul-Fitr"),
+            ("2025-04-18", "Good Friday"),
+            ("2025-04-21", "Easter Monday"),
+            ("2025-05-30", "Indian Arrival Day"),
```
Looks like Labour Day got forgotten here:
```suggestion
            ("2025-04-21", "Easter Monday"),
            ("2025-05-01", "Labour Day"),
            ("2025-05-30", "Indian Arrival Day"),
```

Roshan-1024 replied:
According to [Wiki](https://en.wikipedia.org/wiki/Labour_Day#:~:text=In%20Trinidad%20and%20Tobago%2C%20Labour%20Day%20is%20celebrated%20every%2019%20June.) Labour Day falls on "2025-06-19" for TT.

PPsyrius replied:
Good catch

PPsyrius commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,270 @@
+            ("2025-03-30", "Spiritual Baptist Shouter Liberation Day"),
+            ("2025-03-31", "Eid-Ul-Fitr"),
+            ("2025-04-18", "Good Friday"),
+            ("2025-04-21", "Easter Monday"),
+            ("2025-05-30", "Indian Arrival Day"),
```
```suggestion
            ("2025-04-21", "Easter Monday"),
            ("2025-05-01", "Labor Day"),
            ("2025-05-30", "Indian Arrival Day"),
```

Roshan-1024 replied:
Same goes for this as well.

PPsyrius commented on holidays/locale/en_US/LC_MESSAGES/TT.po:
```diff
@@ -0,0 +1,96 @@
+
+#. Eid al-Fitr.
+msgid "Eid-Ul-Fitr"
+msgstr "Eid-Ul-Fitr"
```
This should be `en_US`:
```suggestion
msgstr "Eid al-Fitr"
```

PPsyrius commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,270 @@
+            ("2025-03-03", "Carnival Monday"),
+            ("2025-03-04", "Carnival Tuesday"),
+            ("2025-03-30", "Spiritual Baptist Shouter Liberation Day"),
+            ("2025-03-31", "Eid-Ul-Fitr"),
```
```suggestion
            ("2025-03-31", "Eid al-Fitr"),
```

PPsyrius commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,186 @@
+        self._add_holiday_jun_19(tr("Labour Day"))
+
+        # African Emancipation Day.
+        self._add_holiday_aug_1(tr("Emancipation Day"))
```
I don't think this get implements yet (remember to update the test cases for this as well)

PPsyrius commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,270 @@
+            "2025-03-04",
+            "2026-02-17",
+        )
+        self.assertHolidayName(name, range(1963, 2027))
```
```suggestion
        self.assertHolidayName(name, range(1963, 2050))
```

PPsyrius commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,270 @@
+            "2025-03-03",
+            "2026-02-16",
+        )
+        self.assertHolidayName(name, range(1963, 2027))
```
```suggestion
        self.assertHolidayName(name, range(1963, 2050))
```

KJhellico commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,159 @@
+        self._add_eid_al_fitr_day(tr("Eid-Ul-Fitr"))
+
+        # Diwali.
+        self._add_diwali(tr("Diwali"))
```
❓

Roshan-1024 replied:
I think I have placed it correctly this time. Please take a look @KJhellico

Roshan-1024 replied:
Could you please take a look now @KJhellico . Let me know if the order is right

KJhellico replied:
https://github.com/vacanza/holidays/pull/2386#issuecomment-2780675157

Roshan-1024 replied:
Hey @KJhellico , 
I calculated the ranges of the movable christian holidays:
- Carnival Monday: Feb 2 - Mar 8
- Carnival Tuesday: Feb 3 - Mar 9 

- Good Friday: Mar 20 – Apr 23
- Easter Monday: Mar 23 – Apr 26
- Corpus Christi: May 21 – Jun 24

So, I have placed the carnival monday and tuesday between new year's day and Spiritual Baptist Liberation Day.
Now for the other 3 holidays, since they don't fall strictly between any two holidays, I'm not sure where to place them. Any suggestions?

Roshan-1024 replied:
Btw, just found out that the first two are optional holidays from [here](https://otp.tt/trinidad-and-tobago/national-holidays-and-awards/). Should I remove them?

Roshan-1024 replied:
Btw, just found out that the first two are optional holidays from [here](https://otp.tt/trinidad-and-tobago/national-holidays-and-awards/). Should I remove them?
@KJhellico @PPsyrius

KJhellico replied:
To keep it simple, place Good Friday and Easter Monday before Spiritual Baptist Liberation Day, and Corpus Christi before Labour Day.

> first two are optional holidays

Yes, I noticed that too. You can move it to OPTIONAL category.

Roshan-1024 replied:
Thanks for the suggestions!

KJhellico commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,160 @@
+        # Labor Day.
+        self._add_holiday_jun_19(tr("Labour Day"))
+
+        # African Emancipation Day.
+        self._add_holiday_aug_1(tr("African Emancipation Day"))
```
[In 1985](http://www.nalis.gov.tt/Research/SubjectGuide/EmancipationDay/tabid/189/Default.aspx).

Roshan-1024 replied:
It shows page not found...

KJhellico replied:
[Use the force](https://web.archive.org/web/20161027070935/https://www.nalis.gov.tt/Research/SubjectGuide/EmancipationDay/tabid/189/Default.aspx) :)

KJhellico commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,160 @@
+        # Independence Day.
+        self._add_holiday_aug_31(tr("Independence Day"))
+
+        # Republic Day.
+        self._add_holiday_sep_24(tr("Republic Day"))
```
Trinidad and Tobago became a republic in 1976.

KJhellico commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,160 @@
+        self._add_holiday_sep_24(tr("Republic Day"))
+
+        # Diwali.
+        self._add_diwali(tr("Diwali"))
```
```suggestion
        self._add_diwali(tr("Divali"))
```
Later we need to unify en_US name for this holiday (for now we can leave "Diwali", I think).

Roshan-1024 replied:
okay!

KJhellico replied:
❓

PPsyrius replied:
```suggestion
        dts_observed.update(self._add_diwali(tr("Divali")))
```

KJhellico commented:
According to[ this source](https://web.archive.org/web/20090221084249/https://www.gov.tt/citizen/service.aspx?id=%7B0CBCBA27-A93B-4834-82B9-AF0734E753B9%7D):
> When a public holiday falls on a Sunday, the holiday will be observed on the Monday immediately following. When two public holidays fall on the same day, the following day is also given as a public holiday.

Roshan-1024 commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,160 @@
+        self._add_new_years_day(tr("New Year's Day"))
+
+        if self._year >= 1996:
+            # Spiritual Baptist Shouter Liberation Day.
+            self._add_holiday_mar_30(tr("Spiritual Baptist Shouter Liberation Day"))
```
https://otp.tt/trinidad-and-tobago/national-holidays-and-awards/

Roshan-1024 commented:
Hey @KJhellico ,
After the latest update (v0.71), I ran `make check` and noticed it modified these files:

1. 	modified:   holidays/locale/be/LC_MESSAGES/BY.po
2. 	modified:   holidays/locale/en_US/LC_MESSAGES/BY.po
3. 	modified:   holidays/locale/ru/LC_MESSAGES/BY.po
4. 	modified:   holidays/locale/th/LC_MESSAGES/BY.po

Since my PR is focused on TT holidays, I think these changes are unrelated.

Roshan-1024 commented:
> Hey @KJhellico , After the latest update (v0.71), I ran `make check` and noticed it modified these files:
> 
> 1. modified:   holidays/locale/be/LC_MESSAGES/BY.po
> 2. modified:   holidays/locale/en_US/LC_MESSAGES/BY.po
> 3. modified:   holidays/locale/ru/LC_MESSAGES/BY.po
> 4. modified:   holidays/locale/th/LC_MESSAGES/BY.po
> 
> Since my PR is focused on TT holidays, I think these changes are unrelated.

I am sharing the git diff here.

```
(venv) (base) roshan@roshan-HP-Pavilion-Gaming-Laptop-15-ec0xxx:~/Desktop/holidays3$ git diff
diff --git a/holidays/locale/be/LC_MESSAGES/BY.po b/holidays/locale/be/LC_MESSAGES/BY.po
index 25408184..b08d9cbc 100644
--- a/holidays/locale/be/LC_MESSAGES/BY.po
+++ b/holidays/locale/be/LC_MESSAGES/BY.po
@@ -48,8 +48,8 @@ msgid "Дзень Перамогі"
 msgstr ""
 
 #. Day of the National Coat of Arms of the Republic of Belarus, the National
-#. Flag of the Republic of
-#. Belarus and the National Anthem of the Republic of Belarus.
+#. Flag of the Republic
+#. of Belarus and the National Anthem of the Republic of Belarus.
diff --git a/holidays/locale/be/LC_MESSAGES/BY.po b/holidays/locale/be/LC_MESSAGES/BY.po
index 25408184..b08d9cbc 100644
--- a/holidays/locale/be/LC_MESSAGES/BY.po
+++ b/holidays/locale/be/LC_MESSAGES/BY.po
@@ -48,8 +48,8 @@ msgid "Дзень Перамогі"
 msgstr ""
 
 #. Day of the National Coat of Arms of the Republic of Belarus, the National
-#. Flag of the Republic of
-#. Belarus and the National Anthem of the Republic of Belarus.
+#. Flag of the Republic
+#. of Belarus and the National Anthem of the Republic of Belarus.
 msgid "Дзень Дзяржаўнага сцяга, Дзяржаўнага герба і Дзяржаўнага гімна Рэспублікі Беларусь"
 msgstr ""
 
diff --git a/holidays/locale/en_US/LC_MESSAGES/BY.po b/holidays/locale/en_US/LC_MESSAGES/BY.po
index 4761b984..b33f50e9 100644
--- a/holidays/locale/en_US/LC_MESSAGES/BY.po
+++ b/holidays/locale/en_US/LC_MESSAGES/BY.po
@@ -48,8 +48,8 @@ msgid "Дзень Перамогі"
 msgstr "Victory Day"
 
 #. Day of the National Coat of Arms of the Republic of Belarus, the National
-#. Flag of the Republic of
-#. Belarus and the National Anthem of the Republic of Belarus.
+#. Flag of the Republic
+#. of Belarus and the National Anthem of the Republic of Belarus.
 msgid "Дзень Дзяржаўнага сцяга, Дзяржаўнага герба і Дзяржаўнага гімна Рэспублікі Беларусь"
 msgstr ""
 "Day of the National Coat of Arms of the Republic of Belarus, the National Flag of the Republic of"
diff --git a/holidays/locale/ru/LC_MESSAGES/BY.po b/holidays/locale/ru/LC_MESSAGES/BY.po
index 309d8750..a48ed4ca 100644
--- a/holidays/locale/ru/LC_MESSAGES/BY.po
+++ b/holidays/locale/ru/LC_MESSAGES/BY.po
@@ -48,8 +48,8 @@ msgid "Дзень Перамогі"
 msgstr "День Победы"
 
 #. Day of the National Coat of Arms of the Republic of Belarus, the National
-#. Flag of the Republic of
-#. Belarus and the National Anthem of the Republic of Belarus.
+#. Flag of the Republic
+#. of Belarus and the National Anthem of the Republic of Belarus.
 msgid "Дзень Дзяржаўнага сцяга, Дзяржаўнага герба і Дзяржаўнага гімна Рэспублікі Беларусь"
 msgstr ""
 "День Государственного флага, Государственного герба и Государственного гимна Республики Беларусь"
diff --git a/holidays/locale/th/LC_MESSAGES/BY.po b/holidays/locale/th/LC_MESSAGES/BY.po
index b29c7a4f..c630b570 100644
--- a/holidays/locale/th/LC_MESSAGES/BY.po
+++ b/holidays/locale/th/LC_MESSAGES/BY.po
@@ -48,8 +48,8 @@ msgid "Дзень Перамогі"
 msgstr "วันแห่งชัยชนะ"
 
 #. Day of the National Coat of Arms of the Republic of Belarus, the National
-#. Flag of the Republic of
-#. Belarus and the National Anthem of the Republic of Belarus.
+#. Flag of the Republic
+#. of Belarus and the National Anthem of the Republic of Belarus.
 msgid "Дзень Дзяржаўнага сцяга, Дзяржаўнага герба і Дзяржаўнага гімна Рэспублікі Беларусь"
 msgstr "วันธงชาติ ตราแผ่นดิน และเพลงชาติแห่งสาธารณรัฐเบลารุส"
 
(END)


```

KJhellico commented:
We work on this problem (#2459), but there's still work to be done.

KJhellico commented:
Problem with .po files solved, you can update your branch.

Roshan-1024 commented:
> Problem with .po files solved, you can update your branch.

Alright!

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,271 @@
+        self.assertHolidayName(
+            name,
+            self.optional_holidays,
+            "2009-02-23",
+            "2010-02-15",
+            "2011-03-07",
+            "2012-02-20",
+            "2013-02-11",
+            "2014-03-03",
+            "2015-02-16",
+            "2016-02-08",
+            "2017-02-27",
+            "2018-02-12",
+            "2019-03-04",
+            "2020-02-24",
+            "2023-02-20",
+            "2024-02-12",
+            "2025-03-03",
+            "2026-02-16",
```
```suggestion
            "2020-02-24",
            "2021-02-15",
            "2022-02-28",
            "2023-02-20",
            "2024-02-12",
            "2025-03-03",
```

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,271 @@
+        self.assertHolidayName(name, self.optional_holidays, range(1963, 2050))
+
+    def test_carnival_tuesday(self):
+        name = "Carnival Tuesday"
```
```suggestion
        name = "Carnival Tuesday"
        self.assertNoHolidayName(name)
```

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,271 @@
+        self.assertHolidayName(
+            name,
+            self.optional_holidays,
+            "2009-02-24",
+            "2010-02-16",
+            "2011-03-08",
+            "2012-02-21",
+            "2013-02-12",
+            "2014-03-04",
+            "2015-02-17",
+            "2016-02-09",
+            "2017-02-28",
+            "2018-02-13",
+            "2019-03-05",
+            "2020-02-25",
+            "2023-02-21",
+            "2024-02-13",
+            "2025-03-04",
+            "2026-02-17",
```
```suggestion
            "2020-02-25",
            "2021-02-16",
            "2022-03-01",
            "2023-02-21",
            "2024-02-13",
            "2025-03-04",
```

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,271 @@
+            ("2025-12-25", "Christmas Day"),
+            ("2025-12-26", "Boxing Day"),
+        )
+
```
It would also be good to add `test_optional` for al least 2025.

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,238 @@
+
+    def test_good_friday(self):
+        self.assertHolidayName(
+            "Good Friday",
```
When a value is used more than once, variable is needed.

> And same for similar holidays.

Similar, not all.

Roshan-1024 replied:
Got it!

KJhellico commented:
So, the remaining questions are:
- Diwali holiday name
- observed holidays

PPsyrius commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,163 @@
+        super().__init__(*args, **kwargs)
+
+    def _populate_public_holidays(self):
+        # New Year's Day.
+        self._add_new_years_day(tr("New Year's Day"))
```
`dts_observed.add( ... )` should be add for any entry that can falls on Sunday
```suggestion
        dts_observed = set()

        # New Year's Day.
        dts_observed.add(self._add_new_years_day(tr("New Year's Day")))
```

PPsyrius commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,163 @@
+        # Boxing Day.
+        self._add_christmas_day_two(tr("Boxing Day"))
+
+        # Eid al-Fitr.
+        self._add_eid_al_fitr_day(tr("Eid-Ul-Fitr"))
```
For movable holidays, we use `dts_observed.update( ... )` instead.
```suggestion
        # Eid al-Fitr.
        dts_observed.update(self._add_eid_al_fitr_day(tr("Eid-Ul-Fitr")))

        if self.observed:
            self._populate_observed(dts_observed)
```

PPsyrius commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,163 @@
+        # Diwali.
+        self._add_diwali(tr("Diwali"))
+
+        # Christmas Day.
+        self._add_christmas_day(tr("Christmas Day"))
+
+        # Boxing Day.
+        self._add_christmas_day_two(tr("Boxing Day"))
```
For consecutive days, `observed_rule` override may be needed here i.e. Christmas ones:
```suggestion
        # Christmas Day.
        dts_observed.add(self._add_christmas_day(tr("Christmas Day")), rule=SUN_TO_NEXT_TUE)

        # Boxing Day.
        dts_observed.add(self._add_christmas_day_two(tr("Boxing Day")))
```

Roshan-1024 replied:
```
       # Christmas Day.
>       dts_observed.add(self._add_christmas_day(tr("Christmas Day")), rule=SUN_TO_NEXT_TUE)
E       TypeError: set.add() takes no keyword arguments
```
@PPsyrius I don't think set.add() accepts a second parameter... Could you clarify?

KJhellico replied:
Yes, `dts_observed`-way don't supports supports only one rule, so seems we need to use `SUN_TO_NEXT_WORKDAY`. And implement support for _"When two public holidays fall on the same day, the following day is also given as a public holiday"_ separately. I think it will be similar to South Korea.

Roshan-1024 replied:
There are two cases we should be considering:
1. When Boxing Day falls on Sunday,
-> Boxing Day is observed on Monday.
2. When Christmas Day falls on Sunday,
-> Christmas Day is observed on Monday. So, Boxing Day is observed on Tuesday.

This means that, Boxing day needs to be shifted from Monday to Tuesday only when Christmas Day falls on Sunday. I think we can do something like this:
`boxing_day_date + timedelta(days=+1)`

So, we can apply SUN_TO_NEXT_MON rule to these holidays:
1. New Years Day
2. Spiritual Baptist Liberation Day
3. Indian Arrival Day
4. Labour Day
5. Emancipation Day
6. Independence Day
7. Republic Day
8. Divali
9. Christmas Day
10. Boxing Day
11. Eid-Ul-Fitr

**Is it correct?**

I have also found that from the year 2000 to 2025, there is one more instance which follows the same pattern:(2025, 3, 30) -> Spiritual Baptist Liberation Day (Sunday)
(2025, 3, 31) -> Eid-Ul-Fitr

Roshan-1024 replied:
Also,I think, applying SUN_TO_NEXT_WORKDAY on Christmas Day (Sunday), will shift it to Tuesday as Monday is Boxing Day. I guess, that's not what we want...

Roshan-1024 replied:
@KJhellico ?

KJhellico replied:
> 2. When Christmas Day falls on Sunday,
>    -> Christmas Day is observed on Monday. So, Boxing Day is observed on Tuesday.

Why exactly _Boxing Day_ is observed on Tuesday? 🙄 I think, it's Christmas Day observed.

Roshan-1024 replied:
@KJhellico https://foreign.gov.tt/documents/716/Public_Holidays_2022.pdf

Roshan-1024 replied:
If Christmas Day shifts to Monday, then boxing day has to move to Tuesday because "When two public holidays fall on the same day, the following day is also given as a public holiday" as you mentioned.

KJhellico replied:
"When two public holidays fall on the same day" - it's like 2025-06-19, when Labour Day and Corpus Christi overlap. Christmas Day/Boxing Day case - it's "next following day that is not a public holiday shall be a public holiday".

Roshan-1024 replied:
@KJhellico , These are the years when Christmas Day fell on Sunday:
1. [2005](https://holidayapi.com/countries/tt-pos/2005)
2. [2011](https://holidayapi.com/countries/tt/2011)
3. [2016](https://holidayapi.com/countries/tt/2016)
4. [2022](https://foreign.gov.tt/documents/716/Public_Holidays_2022.pdf)
And so, Christmas is observed on Monday (26th) and Boxing Day is observed on Tuesday (27th)

KJhellico replied:
OK, you can try to implement this.

PPsyrius replied:
@Roshan-1024 @KJhellico I did look a bit into this and only the foreign affairs (Trinidad & Tobago consulate in Panama) labelled as such others left their names empty
- [2016](https://archives.newsday.co.tt/2016/11/19/eca-announces-public-holidays/)
- [2022](https://trinidadexpress.com/newsextra/workers-get-an-additional-public-holiday-for-christmas/article_213c2b4a-61e6-11ed-81ca-0ba4742cf580.html)

Roshan-1024 replied:
Hey, @PPsyrius @KJhellico 
Just to clarify, should I still proceed with the implementation where Christmas falling on a Sunday leads to Boxing Day being observed on Tuesday? Or would it be better to keep Boxing Day on Monday unless there's clear overlap, like with Labour Day and Corpus Christi?

Roshan-1024 replied:
Is there an official rule that determines which holiday is prioritized when two fall on the same day?

KJhellico replied:
> Or would it be better to keep Boxing Day on Monday

As I said, it will be enough to leave Christmas Day and Boxing Day in their usual places, just implement observed holiday.

> Also, in overlapping cases like Labour Day and Corpus Christi, which holiday do we consider as observed on the next day?

Since there is no clear statement on which of the two holidays is transferred, I think it would be better to transfer both (to one day, of course).

> Is there an official rule that determines which holiday is prioritized when two fall on the same day?

I don't see any such rule.

Roshan-1024 replied:
Hey @KJhellico , I’ve implemented the overlapping holiday observation logic (e.g., Corpus Christi and Labour Day on 2025-06-19 observed on 2025-06-20). Also, I found that this exact rule doesn’t exist in SK... So, please take a look and share any feedback or changes - I’m open to suggestions.

KJhellico replied:
Exact rule - of course, not. But the concept itself...

PPsyrius commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,243 @@
+    @classmethod
+    def setUpClass(cls):
+        years = range(1963, 2050)
+        super().setUpClass(TrinidadAndTobago, years=years)
```
To account for our new Observed ones.
```suggestion
        super().setUpClass(TrinidadAndTobago, years=years, years_non_observed=years)
```

PPsyrius commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,243 @@
+        self.assertNoHolidays(TrinidadAndTobago(years=1962))
+        self.assertNoHolidays(TrinidadAndTobago(categories=OPTIONAL, years=1962))
+
+    def test_new_years_day(self):
+        self.assertHolidayName("New Year's Day", (f"{year}-01-01" for year in range(1963, 2050)))
```
Same goes for other cases (I think the newly added [`test_sierra_leone.py`](https://github.com/vacanza/holidays/blob/0ee2d777ed33a1c3d06181d068eb5564d271c6aa/tests/countries/test_sierra_leone.py#L19)) got some nice examples here:
```suggestion
    def test_new_years_day(self):
        name = "New Year's Day"
        self.assertHolidayName(name, (f"{year}-01-01" for year in range(1963, 2050)))
        dt = (
            "2012-01-02",
            "2017-01-02",
            "2023-01-02",
        )
        self.assertHolidayName(f"{name} (observed)", self.no_estimated_holidays, dt)
        self.assertNoNonObservedHoliday(dt)
```

PPsyrius reviewed (commented):
Aside from "Diwali" -> "Divali" change for `en_TT` that @KJhellico mentioned before, I think this should cover all observance changes remain (+some additional holidays observance histories I managed to dig up)

Roshan-1024 commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,163 @@
+    """Trinidad and Tobago holidays.
+
+    References:
+        * <https://otp.tt/trinidad-and-tobago/national-holidays-and-awards/>
```
Thanks for the detailed research!

Roshan-1024 commented:
In 2006-01-01, New Years Day falls on a Sunday.
In 2012-01-01, New Years Day falls on a Sunday.
In 2017-01-01, New Years Day falls on a Sunday.
In 2023-01-01, New Years Day falls on a Sunday.

In 2003-03-30, Spiritual Baptist Liberation Day falls on a Sunday.
In 2008-03-30, Spiritual Baptist Liberation Day falls on a Sunday.
In 2014-03-30, Spiritual Baptist Liberation Day falls on a Sunday.
In 2025-03-30, Spiritual Baptist Liberation Day falls on a Sunday.

In 2004-05-30, Indian Arrival Day falls on a Sunday.
In 2010-05-30, Indian Arrival Day falls on a Sunday.
In 2021-05-30, Indian Arrival Day falls on a Sunday.

In 2005-06-19, Labour Day falls on a Sunday.
In 2011-06-19, Labour Day falls on a Sunday.
In 2016-06-19, Labour Day falls on a Sunday.
In 2022-06-19, Labour Day falls on a Sunday.

In 2004-08-01, Emancipation Day falls on a Sunday.
In 2010-08-01, Emancipation Day falls on a Sunday.
In 2021-08-01, Emancipation Day falls on a Sunday.
In 2003-08-31, Independence Day falls on a Sunday.
In 2008-08-31, Independence Day falls on a Sunday.
In 2014-08-31, Independence Day falls on a Sunday.
In 2025-08-31, Independence Day falls on a Sunday.

In 2000-09-24, Republic Day falls on a Sunday.
In 2006-09-24, Republic Day falls on a Sunday.
In 2017-09-24, Republic Day falls on a Sunday.
In 2023-09-24, Republic Day falls on a Sunday.

In 2005-12-25, Christmas Day falls on a Sunday.
In 2011-12-25, Christmas Day falls on a Sunday.
In 2016-12-25, Christmas Day falls on a Sunday.
In 2022-12-25, Christmas Day falls on a Sunday.

In 2004-12-26, Boxing Day falls on a Sunday.
In 2010-12-26, Boxing Day falls on a Sunday.
In 2021-12-26, Boxing Day falls on a Sunday.

In 2019‑10‑27, Divali falls on Sunday.
In 2023‑11‑12, Divali falls on Sunday.

In 2012‑08‑19, Eid-Ul-Fitr falls on Sunday.
In 2020‑05‑24, Eid-Ul-Fitr falls on Sunday.


I collected all the holidays that fall on Sunday from year 2000 to 2025. Do you think this will be useful? @KJhellico

KJhellico commented:
> I collected all the holidays that fall on Sunday from year 2000 to 2025. Do you think this will be useful?

It will be useful for tests :)

KJhellico commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,237 @@
+        )
+        kwargs.setdefault("observed_rule", SUN_TO_NEXT_MON)
+        super().__init__(*args, **kwargs)
+
```
```suggestion

    def _populate_observed(self, dts: set[date], multiple: bool = False) -> None:
        for dt in sorted(dts):
            self._add_observed(
                dt,
                rule=WORKDAY_TO_NEXT_WORKDAY + SAT_SUN_TO_NEXT_WORKDAY
                if len(self.get_list(dt)) > 1
                else SUN_TO_NEXT_WORKDAY,
            )

```
😉

Roshan-1024 replied:
great!

KJhellico commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,237 @@
+    InternationalHolidays,
+    IslamicHolidays,
+)
+from holidays.observed_holiday_base import ObservedHolidayBase, SUN_TO_NEXT_MON
```
```suggestion
from holidays.observed_holiday_base import (
    ObservedHolidayBase,
    SUN_TO_NEXT_WORKDAY,
    SAT_SUN_TO_NEXT_WORKDAY,
    WORKDAY_TO_NEXT_WORKDAY,
)
```

KJhellico commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,197 @@
+        dts_observed.add(self._add_new_years_day(tr("New Year's Day")))
+
+        # Good Friday.
+        self._add_good_friday(tr("Good Friday"))
```
```suggestion
        dts_observed.add(self._add_good_friday(tr("Good Friday")))
```
Also Easter Monday and Corpus Christi, for completeness.

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,333 @@
+            "2017-01-02",
+            "2023-01-02",
+        )
+        self.assertHolidayName(f"{name} (observed)", self.no_estimated_holidays, dt)
```
```suggestion
        self.assertHolidayName(f"{name} (observed)", dt)
```
We use `self.no_estimated_holidays` to check only holidays that may have an "estimated" label.

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,333 @@
+            "2025-04-18",
+        )
+        self.assertHolidayName(name, range(1963, 2050))
+
```
"Observed" dates test needed for all holidays.

KJhellico replied:
_All_ holidays.

Roshan-1024 replied:
Hi @KJhellico,
I just wanted to let you know that my college exams have recently started, which might have slightly affected the quality of my recent contributions. I sincerely apologize if my work seemed a bit haphazard...
However, I want to assure you that during the GSoC period, I will be completely free from academic commitments and will be able to fully dedicate myself to the project.

Thanks again for your understanding!

KJhellico replied:
You're almost done with this PR, the final touches are left. :)

Roshan-1024 replied:
I will try to complete the remaining work asap.

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,333 @@
+        dt = (
+            "2003-03-31",
+            "2008-03-31",
+            "2014-03-31",
```
```suggestion
            "2014-03-31",
            "2018-04-03",
```

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,333 @@
+        dt = (
+            "2004-05-31",
+            "2010-05-31",
+            "2021-05-31",
```
```suggestion
            "2013-05-31",
            "2021-05-31",
            "2024-05-31",
```

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,333 @@
+        dt = (
+            "2005-06-20",
+            "2011-06-20",
+            "2016-06-20",
+            "2022-06-20",
```
```suggestion
            "2014-06-20",
            "2016-06-20",
            "2022-06-20",
            "2025-06-20",
```

KJhellico commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,197 @@
+        )
+        super().__init__(*args, **kwargs)
+
+    def _populate_observed(self, dts: set[date], multiple: bool = False) -> None:
+        for dt in sorted(dts):
+            self._add_observed(
+                dt,
+                rule=WORKDAY_TO_NEXT_WORKDAY + SAT_SUN_TO_NEXT_WORKDAY
+                if len(self.get_list(dt)) > 1
+                else SUN_TO_NEXT_WORKDAY,
+            )
```
@coderabbitai, This is method signature from the parent class.

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,358 @@
+            "2024-03-29",
+            "2025-04-18",
+        )
+        dt = ("2018-04-03",)
+        self.assertHolidayName(f"{name} (observed)", dt)
+        self.assertHolidayName(name, range(1963, 2050))
```
```suggestion
        self.assertHolidayName(name, range(1963, 2050))
        dt = (
            "2018-04-03",
            "2029-04-03",
        )
        self.assertHolidayName(f"{name} (observed)", dt)
        self.assertNoNonObservedHoliday(dt)
```
`self.assertHolidayName(name, ...` is a part of general holiday checks, it should be after them, not after observed holidays checks. (Update this for rest of holidays)

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,358 @@
+            "2024-04-01",
+            "2025-04-21",
+        )
+        self.assertHolidayName(name, range(1963, 2050))
```
```suggestion
        self.assertHolidayName(name, range(1963, 2050))
        dt = (
            "2043-03-31",
            "2054-03-31",
        )
        self.assertHolidayName(f"{name} (observed)", dt)
        self.assertNoNonObservedHoliday(dt)
```
> "Observed" dates test needed for all holidays.

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,358 @@
+            "2018-04-03",
+            "2025-04-01",
+        )
+        self.assertHolidayName(f"{name} (observed)", dt)
```
```suggestion
        self.assertHolidayName(f"{name} (observed)", dt)
        self.assertNoNonObservedHoliday(dt)
```

KJhellico commented on tests/countries/test_trinidad_and_tobago.py:
```diff
@@ -0,0 +1,358 @@
+            "2016-12-27",
+            "2022-12-27",
+        )
+        self.assertHolidayName(f"{name} (observed)", dt)
```
```suggestion
        self.assertHolidayName(f"{name} (observed)", dt)
        self.assertNoNonObservedHoliday(dt)
```

Roshan-1024 commented:
@KJhellico @PPsyrius Please let me know if there are any other changes you'd like me to make.

KJhellico reviewed (approved):
LGTM! 👍

PPsyrius commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,197 @@
+    """Trinidad and Tobago holidays.
+
+    References:
+        * <https://otp.tt/trinidad-and-tobago/national-holidays-and-awards/>
+        * <https://laws.gov.tt/ttdll-web2/revision/list?offset=380&q=&currentid=322>
+        * <https://www.guardian.co.tt/article-6.2.448640.3e9e366923>
+        * <https://www.guardian.co.tt/article-6.2.410311.3fe66fb00f>
+        * <https://www.nalis.gov.tt/resources/tt-content-guide/labour-day/>
+        * <https://www.facebook.com/groups/191766699268/posts/10160832951274269/>
+        * <https://www.facebook.com/100064996051675/posts/1101755938667598/?_rdr>
+        * <https://en.wikipedia.org/wiki/Public_holidays_in_Trinidad_and_Tobago>
+        * <https://en.wikipedia.org/wiki/Trinidad_and_Tobago_Carnival>
+        * <https://en.wikipedia.org/wiki/Indian_Arrival_Day#Trinidad_and_Tobago>
+        * <https://www.timeanddate.com/holidays/trinidad/eid-al-fitr>
+        * <https://www.timeanddate.com/holidays/trinidad/diwali>
```
Let's archive our sources as well to make this fully compatible for #2504 
```suggestion
        * <https://web.archive.org/web/20250429121623/https://otp.tt/trinidad-and-tobago/national-holidays-and-awards/>
        * <https://web.archive.org/web/20250429120744/https://laws.gov.tt/ttdll-web2/revision/list?offset=380&q=&currentid=322>
        * <https://web.archive.org/web/20250429120707/https://www.guardian.co.tt/article-6.2.448640.3e9e366923>
        * <https://web.archive.org/web/20250429120548/https://www.guardian.co.tt/article-6.2.410311.3fe66fb00f>
        * <http://archive.today/2025.04.29-121801/https://www.nalis.gov.tt/resources/tt-content-guide/labour-day/>
        * <https://web.archive.org/web/20250429120701/https://www.facebook.com/groups/191766699268/posts/10160832951274269/>
        * <https://web.archive.org/save/https://www.facebook.com/100064996051675/posts/1101755938667598/>
        * <https://en.wikipedia.org/wiki/Public_holidays_in_Trinidad_and_Tobago>
        * <https://en.wikipedia.org/wiki/Trinidad_and_Tobago_Carnival>
        * <https://en.wikipedia.org/wiki/Indian_Arrival_Day#Trinidad_and_Tobago>
        * <https://web.archive.org/web/20250429120402/https://www.timeanddate.com/holidays/trinidad/eid-al-fitr>
        * <https://web.archive.org/web/20250429120140/https://www.timeanddate.com/holidays/trinidad/diwali>
```

PPsyrius commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,197 @@
+        * <https://web.archive.org/web/20250429120548/https://www.guardian.co.tt/article-6.2.410311.3fe66fb00f>
+        * <http://archive.today/2025.04.29-121801/https://www.nalis.gov.tt/resources/tt-content-guide/labour-day/>
+        * <https://web.archive.org/web/20250429120701/https://www.facebook.com/groups/191766699268/posts/10160832951274269/>
+        * <https://web.archive.org/save/https://www.facebook.com/100064996051675/posts/1101755938667598/>
```
Sorry about the wrong link earlier 😔 
```suggestion
        * <https://web.archive.org/web/20250429142054/https://www.facebook.com/plugins/post.php?href=https%3A%2F%2Fwww.facebook.com%2F100064996051675%2Fposts%2F1101755938667598%2F>
```

KJhellico commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,197 @@
+        * <https://web.archive.org/web/20250429120744/https://laws.gov.tt/ttdll-web2/revision/list?offset=380&q=&currentid=322>
+        * <https://web.archive.org/web/20250429120707/https://www.guardian.co.tt/article-6.2.448640.3e9e366923>
+        * <https://web.archive.org/web/20250429120548/https://www.guardian.co.tt/article-6.2.410311.3fe66fb00f>
+        * <http://archive.today/2025.04.29-121801/https://www.nalis.gov.tt/resources/tt-content-guide/labour-day/>
```
[Wayback Machine snapshot](https://web.archive.org/web/20250418212357/https://www.nalis.gov.tt/resources/tt-content-guide/labour-day/) is also working.

PPsyrius replied:
Nice, I tried it earlier but it keeps failing when I did it for that link for some reason

PPsyrius replied:
```suggestion
        * <https://web.archive.org/web/20250418212357/https://www.nalis.gov.tt/resources/tt-content-guide/labour-day/>
```

PPsyrius reviewed (approved):
LGTM 🇹🇹

KJhellico commented on holidays/countries/trinidad_and_tobago.py:
```diff
@@ -0,0 +1,197 @@
+            # Republic Day.
+            dts_observed.add(self._add_holiday_sep_24(tr("Republic Day")))
+
+        # Diwali.
+        dts_observed.add(self._add_diwali(tr("Divali")))
+
+        # Christmas Day.
+        dts_observed.add(self._add_christmas_day(tr("Christmas Day")))
+
+        # Boxing Day.
+        dts_observed.add(self._add_christmas_day_two(tr("Boxing Day")))
+
+        # Eid al-Fitr.
+        dts_observed.update(self._add_eid_al_fitr_day(tr("Eid-Ul-Fitr")))
```
@coderabbitai, `holidays.calendars.hindu._HinduLunisolar` provides Diwali dates for 1901-2100 years; if `holidays.groups.islamic.IslamicHolidays::_add_eid_al_fitr_day()` return empty set, this will not cause an error.

arkid15r reviewed (commented):
Thank you and congrats on another great contribution @Roshan-1024 !
